### PR TITLE
Fix mirroring from upstream to downstream with ranges

### LINKED ISSUE #23855: Mirroring ranged versions removes ranges for most browsers
### What type of issue is this?

Infrastructure issue

### What is the issue?

Both Chrome and Firefox use ranges in this feature:

https://github.com/mdn/browser-compat-data/blob/9832a6516bfe3d8b1d18e503645d0332267c43a0/html/elements/iframe.json#L767-L800

The data we publish (after mirroring) is this:

```
          "allow-forms": {
            "__compat": {
              "description": "<code>sandbox=\"allow-forms\"</code>",
              "source_file": "html/elements/iframe.json",
              "spec_url": "https://html.spec.whatwg.org/multipage/browsers.html#attr-iframe-sandbox-allow-forms",
              "status": {
                "deprecated": false,
                "experimental": false,
                "standard_track": true
              },
              "support": {
                "chrome": {
                  "version_added": "≤49"
                },
                "chrome_android": {
                  "version_added": "49"
                },
                "edge": {
                  "version_added": "79"
                },
                "firefox": {
                  "version_added": "≤49"
                },
                "firefox_android": {
                  "version_added": "49"
                },
                "ie": {
                  "version_added": false
                },
                "oculus": {
                  "version_added": "5.0"
                },
                "opera": {
                  "version_added": "36"
                },
                "opera_android": {
                  "version_added": "36"
                },
                "safari": {
                  "version_added": "≤10.1"
                },
                "safari_ios": {
                  "version_added": "≤10.3"
                },
                "samsunginternet_android": {
                  "version_added": "5.0"
                },
                "webview_android": {
                  "version_added": "49"
                }
              }
            }
          },
```

Note that there are no ranges for any mirrored data except safari_ios, which is done correctly.

### What behavior were you expecting?

Ranges should be preserved for browsers that mirror from Chrome and Firefox.

### What version(s) of BCD is the issue present in?

- [X] The current BCD release
- [X] The current version of the `main` branch

### Do you have anything more you want to share?

cc @queengooborg

### PR DESCRIPTION
**Pull request by queengooborg**
This PR fixes the mirroring script when ranges are involved.  Apparently, if the upstream release was a range, the downstream browser would not be a range, which meant that various bits of data was not set to mirror when it should.  This PR fixes this by introducing some logic that checks if the previous release (before the matching release) uses the same engine as the upstream browser, and if the upstream has a range, it adds the range delimiter.

The engine check for the last release is to help remove range delimiters where there not needed.  For example, if the resulting version is Edge 79, we know that it's pointless to make it a range ("≤79") since Edge 18 and earlier used a different browser engine.  As such, this change now produces the following:
- Chrome 80 -> Edge 80
- Chrome ≤80 -> Edge ≤80
- Chrome 79 -> Edge 79
- Chrome ≤79 -> Edge 79
- Chrome 24 -> Edge 79
- Chrome ≤24 -> Edge 79

Caveat: due to this change, some data can no longer be mirrored for Firefox Android and Opera Android.  This is a minor caveat, however, since they'll be automatically fixed once the ranges are replaced with proper version numbers.

Fixes #23855 (closes #23856).

### CODE REVIEW

**Comment by Elchi3** [comment]
This PR needs to be reviewed with the approach proposed in https://github.com/mdn/browser-compat-data/pull/23856.

**Comment by queengooborg** [comment]
Ah, I forgot there was an existing PR that also did this, or at least something similar.

Since Philip is now away and won't be back for a long while, I think that we should close that PR and combine its changes into this one.  I say this because I'd request changes on some of the code for both bug fixes and improved logic.

One key difference between that PR and this one is that it bases whether to add the range delimiter if the downstream version is the first release, but this PR only adds the range delimiter if the downstream version is the first release _of the matching engine version,_ which means that less ranges are introduced during mirroring.  (In Philip's PR, Chrome 23 would still become Edge ≤79 during mirroring.)

File: scripts/build/mirror.ts [hunk 1]
```diff
@@ -89,6 +89,8 @@ export const getMatchingBrowserVersion = (
     );
   }
 
+  let lastRelease;
```
Comment (caugner) [inline scripts/build/mirror.ts:92]: To make this a bit clearer:

```suggestion
  let previousReleaseEngine;
```

File: scripts/build/mirror.ts [hunk 2]
```diff
@@ -105,8 +107,15 @@ export const getMatchingBrowserVersion = (
...
     ) {
+      if (range && lastRelease?.engine == sourceRelease.engine) {
+        // Add a range delimiter if there were previous releases of the downstream browser that used the same engine before this one (ex. after Edge 79)
+        return `≤${r}`;
+      }
+
       return r;
     }
+
+    lastRelease = release;
```
Comment (caugner) [inline scripts/build/mirror.ts:118]:
```suggestion
    previousReleaseEngine = release.engine;
```

File: scripts/build/mirror.ts [hunk 3]
```diff
@@ -105,8 +107,15 @@ export const getMatchingBrowserVersion = (
       sourceRelease.engine_version &&
       compare(release.engine_version, sourceRelease.engine_version, '>=')
     ) {
+      if (range && lastRelease?.engine == sourceRelease.engine) {
```
Comment (caugner) [inline scripts/build/mirror.ts:110]:
```suggestion
      if (range && previousReleaseEngine == sourceRelease.engine) {
```

**Review by caugner — CHANGES_REQUESTED** [review]
LGTM, just three comments to make the implementation a bit clearer. And could we add a unit test?

**Comment by queengooborg** [comment]
Since the changes requested have been applied, I'm going to go ahead and merge this since an LGTM was given!

**Comment by sherephonsung** [comment]
> This PR fixes the mirroring script when ranges are involved. Apparently, if the upstream release was a range, the downstream browser would not be a range, which meant that various bits of data was not set to mirror when it should. This PR fixes this by introducing some logic that checks if the previous release (before the matching release) uses the same engine as the upstream browser, and if the upstream has a range, it adds the range delimiter.
> 
> The engine check for the last release is to help remove range delimiters where there not needed. For example, if the resulting version is Edge 79, we know that it's pointless to make it a range ("≤79") since Edge 18 and earlier used a different browser engine. As such, this change now produces the following:
> 
> * Chrome 80 -> Edge 80
> * Chrome ≤80 -> Edge ≤80
> * Chrome 79 -> Edge 79
> * Chrome ≤79 -> Edge 79
> * Chrome 24 -> Edge 79
> * Chrome ≤24 -> Edge 79
> 
> Caveat: due to this change, some data can no longer be mirrored for Firefox Android and Opera Android. This is a minor caveat, however, since they'll be automatically fixed once the ranges are replaced with proper version numbers.
> 
> Fixes #23855 (closes #23856).

#28850animation.overallprograss/
#git_commit

**Pull request merged**